xtask/src/logging.rs

```diff
@@ -1,65 +1,100 @@
     }
-    std::fs::File::create(&log_path)?;
-    Ok(log_path)
+
+    std::fs::File::create(&log_path)
+        .map_err(|e| warn!("Impossible redirect logs, using stdout instead. Error: {e}"))
+        .ok()?;
+    Some(log_path.to_string_lossy().to_string())
 }
 
-// If the log file cannot be created due to any reasons,
-// such as lack of permission to create files or new folders in the path,
-// things will be printed to stdout instead of being redirected to the logs file
+// This macro will accept a description of the command, an xshell::Shell, a command
+// (a string on which format will be called), the log_path, and optionally the process group
+// used to create the command.
 //
-// The returned closure will accept a description of the command and the command itself as a duct::Expression.
-// The description will be printed to both stdout and the log file, if possible, while
-// to the expression will be added the redirection of the logs, if possible.
-pub fn create_with_logs(
-    project_path: &Path,
-    log_path: String,
-) -> Box<dyn Fn(&str, duct::Expression) -> duct::Expression> {
-    let without_logs = |description: &str, cmd: duct::Expression| -> duct::Expression {
-        info!("{description}");
-        cmd
-    };
-
-    let log_path = match create_log_file(project_path, &log_path) {
-        Ok(log_path) => log_path,
-        Err(e) => {
-            warn!("Impossible redirect logs, using stdout instead. Error: {e}");
-            return Box::new(without_logs);
-        }
-    };
+// The description will be logged with the info log level, and the command will be created
+// using the provided shell and converted to a tokio::process::Command to redirect stdout and stderr.
+//
+// If log is None, then things will be redirected to stdout.
+// Instead, if it contains a path, it will be used to redirect the command's output.
+//
+// The provided path needs to be a valid path, `create_log_file` should be used before use this macro
+#[macro_export]
+macro_rules! cmd_with_logs {
+    ($description:expr, $sh:expr, $cmd:literal, $log:expr $(,$gpid:literal)?) => {{
+        use std::process::Stdio;
+        tracing::info!("{}", $description);
+        let (stdout, stderr) = match $log {
+            None => (Stdio::inherit(), Stdio::inherit()),
+            Some(ref log_path) => {
+                // redirecting stdout and stderr into log_path
+                let mut log_out_file = std::fs::File::options()
+                    .append(true)
+                    .create(true)
+                    .open(&log_path)
+                    .expect("Log file does not exist");
+                let log_err_file = log_out_file.try_clone()?;
 
-    let with_logs = move |description: &str, cmd: duct::Expression| -> duct::Expression {
-        // The file has just been created
-        let mut log_file = std::fs::File::options()
-            .append(true)
-            .open(&log_path)
-            .unwrap();
+                use std::io::Write;
```

Comment:
import in the middle of a block is bad style IMO

Reply:
What do you mean by `as _`? Can it be done only on generics, not on variables?

For now I change it to 
```rust
                let _ = std::io::Write::write(&mut log_out_file, format!("{}\n", $description).as_bytes())
                    .map_err(|e| tracing::warn!("Error writing into {log_path}, error: {e}"));

```

Reply:
Just to be sure, is it also considered a bad style to import at the beginning of a block? If multiple calls within the block use that trait, it may be cleaner to have just one import statement and then make all calls without the need to qualify each of them (In a macro context I mean, in standard code it could be moved to the beginning of the file of course)

Reply:
> Nit: as _ would be slightly more better.

as in `use std::io::Write as _`. This is useful to indicate that the trait doesn't appear anywhere explicitly, but rather, that some functions are being used from that trait.